File: ui-tests/PageObjects/Panes/Editor/TraceLogPanel.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Playwright;
using UiTests.PageObjects.Components;
using UiTests.PageObjects.Panes.EventLog;

namespace UiTests.PageObjects.Panes.Editor;

/// <summary>
/// Represents the trace log panel associated with a specific editor line.
/// </summary>
public class TraceLogPanel
{
    public TraceLogPanel(EditorPane parentPane, int lineNumber)
    {
        ParentPane = parentPane;
        LineNumber = lineNumber;
    }

    /// <summary>
    /// Parent editor pane where the trace panel originated.
    /// </summary>
    public EditorPane ParentPane { get; }

    /// <summary>
    /// Line number that owns this trace log panel.
    /// </summary>
    public int LineNumber { get; }

    /// <summary>
    /// Root locator of the trace log panel element.
    /// </summary>
    public ILocator Root
        => ParentPane.Root.Locator($"xpath=//*[@id='edit-trace-{ParentPane.IdNumber}-{LineNumber}']/ancestor::*[@class='trace']");

    /// <summary>
    /// Monaco editor's visible text area where code is rendered.
    /// This is the clickable area that focuses the editor.
    /// </summary>
    public ILocator MonacoViewLines()
        => Root.Locator(".monaco-editor .view-lines").First;

    /// <summary>
    /// Editable text box used to configure the trace expression.
    /// Monaco editor creates a textarea for input. The class name varies by Monaco version:
    /// - Older versions: 'inputarea'
    /// - Newer versions: 'ime-text-area'
    /// We use First to handle any duplicates.
    /// </summary>
    public ILocator EditTextBox()
        => Root.Locator("textarea.inputarea, textarea.ime-text-area").First;

    /// <summary>
    /// Types text into the trace expression editor by clicking the Monaco view area
    /// and using keyboard input. This is the recommended approach for newer Monaco versions.
    /// </summary>
    /// <param name="expression">The trace expression to type.</param>
    public async Task TypeExpressionAsync(string expression)
    {
        var viewLines = MonacoViewLines();
        await viewLines.WaitForAsync(new() { State = WaitForSelectorState.Visible, Timeout = 5000 });

        // Click on the view lines to focus the Monaco editor
        await viewLines.ClickAsync();
        await Task.Delay(200);

        var editId = $"edit-trace-{ParentPane.IdNumber}-{LineNumber}";
        var page = ParentPane.Root.Page;

        // Try to set value via Monaco API - access through data.services or global monaco
        var setViaApi = await page.EvaluateAsync<bool>(@"(args) => {
            const editId = args.editId;
            const expression = args.expression;
            const editDiv = document.getElementById(editId);
            if (!editDiv) return false;

            // Try multiple ways to access Monaco editors
            // Method 1: Via global monaco API
            const monacoEditors = window.monaco?.editor?.getEditors?.() || [];
            for (const editor of monacoEditors) {
                const domNode = editor.getDomNode();
                if (domNode && editDiv.contains(domNode)) {
                    editor.setValue(expression);
                    return true;
                }
            }

            // Method 2: Look for Monaco container's editor property
            const monacoContainer = editDiv.querySelector('.monaco-editor');
            if (monacoContainer && monacoContainer._editorInstance) {
                monacoContainer._editorInstance.setValue(expression);
                return true;
            }

            return false;
        }", new { editId, expression });

        // Always use keyboard fallback for reliability
        // Click to ensure focus
        await viewLines.ClickAsync();
        await Task.Delay(150);

        // Select all and delete existing content
        await page.Keyboard.PressAsync("Control+a");
        await Task.Delay(50);
        await page.Keyboard.PressAsync("Delete");
        await Task.Delay(50);

        // Type the expression character by character with delay for reliability
        await page.Keyboard.TypeAsync(expression, new() { Delay = 30 });
    }

    /// <summary>
    /// Rows rendered in the trace log panel.
    /// </summary>
    /// <remarks>
    /// The trace table can be rendered in different views:
    /// - .trace-view: The default view with trace rows
    /// - .chart-table .trace-table: Alternative table view with DataTables styling
    /// We check both selectors to handle view mode variations.
    /// </remarks>
    public async Task<IReadOnlyList<TraceLogRow>> TraceRowsAsync()
    {
        // Try the chart-table selector first (DataTables-based view)
        var locators = await Root.Locator(".chart-table .trace-table tbody tr").AllAsync();
        if (locators.Count == 0)
        {
            // Fall back to trace-view selector
            locators = await Root.Locator(".trace-view tbody tr").AllAsync();
        }

        var menu = new ContextMenu(ParentPane.Root.Page);
        return locators.Select(l => new TraceLogRow(l, menu)).ToList();
    }

    /// <summary>
    /// Backwards-compatible accessor returning the legacy event row wrapper.
    /// </summary>
    public async Task<IReadOnlyList<EventRow>> EventRowsAsync()
    {
        var traceRows = await TraceRowsAsync();
        return traceRows
            .Select(row => new EventRow(row.Root, EventElementType.TracePointEditor))
            .ToList();
    }

    /// <summary>
    /// The hamburger menu button that opens the dropdown containing Disable/Hide/Delete options.
    /// </summary>
    public ILocator HamburgerMenu() => Root.Locator(".hamburger-dropdown");

    /// <summary>
    /// The dropdown list container that appears when the hamburger menu is clicked.
    /// </summary>
    public ILocator DropdownList() => Root.Locator(".dropdown-list");

    /// <summary>
    /// The Disable/Enable toggle button inside the hamburger dropdown.
    /// Note: This button is only visible after opening the hamburger menu.
    /// </summary>
    public ILocator ToggleButton() => Root.Locator(".trace-disable");

    /// <summary>
    /// Opens the hamburger menu and clicks the Disable/Enable toggle button.
    /// This method handles the dropdown visibility correctly.
    /// The dropdown uses a blur handler that closes it when focus leaves the hamburger button,
    /// so we need to use JavaScript evaluation to click the toggle button reliably.
    /// </summary>
    public async Task ClickToggleButtonAsync()
    {
        var page = ParentPane.Root.Page;
        var editTraceId = $"edit-trace-{ParentPane.IdNumber}-{LineNumber}";

        // Use JavaScript to reliably toggle the disable state
        // This avoids race conditions with blur handlers
        // We find the trace panel by locating the edit-trace element and going up to its ancestor .trace
        await page.EvaluateAsync(@"(editTraceId) => {
            const editTrace = document.getElementById(editTraceId);
            if (!editTrace) {
                console.log('Could not find edit trace element:', editTraceId);
                return;
            }

            // Find the ancestor .trace element
            const trace = editTrace.closest('.trace');
            if (!trace) {
                console.log('Could not find ancestor .trace element');
                return;
            }

            // Find and click the hamburger to open dropdown
            const hamburger = trace.querySelector('.hamburger-dropdown');
            if (hamburger) {
                hamburger.click();

                // Small delay then click the disable button
                setTimeout(() => {
                    const toggleBtn = trace.querySelector('.trace-disable');
                    if (toggleBtn) {
                        toggleBtn.click();
                    }
                }, 150);
            }
        }", editTraceId);

        // Wait for the action to complete
        await Task.Delay(400);
    }

    public ILocator DisabledOverlay() => Root.Locator(".trace-disabled-overlay");
    public ILocator RunButton() => Root.Locator(".trace-run-button-svg").Nth(0);
}

File: src/tests/gui/page-objects/panes/editor/trace-log-panel.ts
import type { Locator, Page } from "@playwright/test";
import type { EditorPane } from "./editor-pane";
import { TraceLogRow } from "./trace-log-row";
import { ContextMenu } from "../../components/context-menu";

const SCROLL_TIMEOUT_MS = 3_000;
const VIEW_LINES_TIMEOUT_MS = 3_000;
const POST_SCROLL_DELAY_MS = 300;
const POST_CLICK_DELAY_MS = 150;
const POST_SELECT_ALL_DELAY_MS = 50;
const POST_TOGGLE_DELAY_MS = 400;
const TYPE_DELAY_MS = 30;

/**
 * Represents the trace log panel associated with a specific editor line.
 *
 * Port of ui-tests/PageObjects/Panes/Editor/TraceLogPanel.cs
 */
export class TraceLogPanel {
  readonly parentPane: EditorPane;
  readonly lineNumber: number;

  constructor(parentPane: EditorPane, lineNumber: number) {
    this.parentPane = parentPane;
    this.lineNumber = lineNumber;
  }

  private get page(): Page {
    return this.parentPane.page;
  }

  get root(): Locator {
    return this.parentPane.root.locator(
      `xpath=//*[@id='edit-trace-${this.parentPane.idNumber}-${this.lineNumber}']/ancestor::*[@class='trace']`,
    );
  }

  monacoViewLines(): Locator {
    return this.root.locator(".monaco-editor .view-lines").first();
  }

  editTextBox(): Locator {
    return this.root.locator("textarea.inputarea, textarea.ime-text-area").first();
  }

  /**
   * Types text into the trace expression editor.
   * Tries Monaco API first, falls back to keyboard input.
   */
  async typeExpression(expression: string): Promise<void> {
    try {
      await this.root.scrollIntoViewIfNeeded({ timeout: SCROLL_TIMEOUT_MS });
      await sleep(POST_SCROLL_DELAY_MS);
    } catch {
      await sleep(POST_SCROLL_DELAY_MS);
    }

    const editId = `edit-trace-${this.parentPane.idNumber}-${this.lineNumber}`;

    // Try Monaco API first
    const setViaApi = await this.page.evaluate(
      ({ editId: eid, expression: expr }) => {
        const editDiv = document.getElementById(eid);
        if (!editDiv) return false;

        // Via global monaco API
        const w = window as any; // eslint-disable-line @typescript-eslint/no-explicit-any
        const monacoEditors = w.monaco?.editor?.getEditors?.() || [];
        for (const editor of monacoEditors) {
          const domNode = editor.getDomNode();
          if (domNode && editDiv.contains(domNode)) {
            editor.setValue(expr);
            return true;
          }
        }

        // Via Monaco container's editor property
        const monacoContainer = editDiv.querySelector(".monaco-editor") as any; // eslint-disable-line @typescript-eslint/no-explicit-any
        if (monacoContainer?._editorInstance) {
          monacoContainer._editorInstance.setValue(expr);
          return true;
        }

        return false;
      },
      { editId, expression },
    );

    if (setViaApi) {
      return;
    }

    // Keyboard fallback
    const viewLines = this.monacoViewLines();
    try {
      await viewLines.waitFor({ state: "visible", timeout: VIEW_LINES_TIMEOUT_MS });
      await viewLines.click();
    } catch {
      await viewLines.click({ force: true });
    }
    await sleep(POST_CLICK_DELAY_MS);

    await this.page.keyboard.press("Control+a");
    await sleep(POST_SELECT_ALL_DELAY_MS);
    await this.page.keyboard.press("Delete");
    await sleep(POST_SELECT_ALL_DELAY_MS);

    await this.page.keyboard.type(expression, { delay: TYPE_DELAY_MS });
  }

  /**
   * Rows rendered in the trace log panel.
   */
  async traceRows(): Promise<TraceLogRow[]> {
    let locators = await this.root.locator(".chart-table .trace-table tbody tr").all();
    if (locators.length === 0) {
      locators = await this.root.locator(".trace-view tbody tr").all();
    }
    const menu = new ContextMenu(this.page);
    return locators.map((l) => new TraceLogRow(l, menu));
  }

  hamburgerMenu(): Locator {
    return this.root.locator(".hamburger-dropdown");
  }

  dropdownList(): Locator {
    return this.root.locator(".dropdown-list");
  }

  toggleButton(): Locator {
    return this.root.locator(".trace-disable");
  }

  /**
   * Opens the hamburger menu and clicks the Disable/Enable toggle button.
   * Uses JavaScript to avoid race conditions with blur handlers.
   */
  async clickToggleButton(): Promise<void> {
    const editTraceId = `edit-trace-${this.parentPane.idNumber}-${this.lineNumber}`;

    await this.page.evaluate((eid: string) => {
      const editTrace = document.getElementById(eid);
      if (!editTrace) return;

      const trace = editTrace.closest(".trace");
      if (!trace) return;

      const hamburger = trace.querySelector(".hamburger-dropdown") as HTMLElement | null;
      if (hamburger) {
        hamburger.click();
        setTimeout(() => {
          const toggleBtn = trace.querySelector(".trace-disable") as HTMLElement | null;
          if (toggleBtn) {
            toggleBtn.click();
          }
        }, 150);
      }
    }, editTraceId);

    await sleep(POST_TOGGLE_DELAY_MS);
  }

  disabledOverlay(): Locator {
    return this.root.locator(".trace-disabled-overlay");
  }

  runButton(): Locator {
    return this.root.locator(".trace-run-button-svg").nth(0);
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
